bcda/cli.go

```diff
@@ -118,94 +131,87 @@ func importCCLF0(fileMetadata cclfFileMetadata) (map[string]cclfFileValidator, e
-		log.Error(err)
 		return err
 	}
 
-	cclf8File := models.CCLFFile{
-		CCLFNum:         8,
-		Name:            fileMetadata.name,
-		ACOCMSID:        fileMetadata.acoID,
-		Timestamp:       fileMetadata.timestamp,
-		PerformanceYear: fileMetadata.perfYear,
-	}
+	return nil
+}
 
-	db := database.GetGORMDbConnection()
-	defer database.Close(db)
+func importCCLF9(fileMetadata cclfFileMetadata) error {
+	err := importCCLF(fileMetadata, func(fileID uint, b []byte, db *gorm.DB) error {
+		const (
+			currIDStart, currIDEnd               = 1, 12
+			prevIDStart, prevIDEnd               = 12, 23
+			prevIDEffDateStart, prevIDEffDateEnd = 23, 33
+			prevIDObsDateStart, prevIDObsDateEnd = 33, 43
+		)
+
+		cclf9 := models.CCLFBeneficiaryXref{
+			FileID:        fileID,
+			XrefIndicator: string(b[0:1]),
+			CurrentNum:    string(b[currIDStart:currIDEnd]),
+			PrevNum:       string(b[prevIDStart:prevIDEnd]),
+			PrevsEfctDt:   string(b[prevIDEffDateStart:prevIDEffDateEnd]),
+			PrevsObsltDt:  string(b[prevIDObsDateStart:prevIDObsDateEnd]),
+		}
+		err := db.Create(&cclf9).Error
+		if err != nil {
+			return errors.Wrap(err, "could not create CCLF9 cross reference record")
+		}
+		return nil
+	})
 
-	err = db.Create(&cclf8File).Error
 	if err != nil {
-		return errors.Wrap(err, "could not create CCLF8 file record")
-	}
-
-	const (
-		mbiStart, mbiEnd   = 0, 11
-		hicnStart, hicnEnd = 11, 22
-	)
-
-	sc := bufio.NewScanner(file)
-	for sc.Scan() {
-		b := sc.Bytes()
-		if len(bytes.TrimSpace(b)) > 0 {
-			err = db.Create(&models.CCLFBeneficiary{
-				FileID: cclf8File.ID,
-				MBI:    string(bytes.TrimSpace(b[mbiStart:mbiEnd])),
-				HICN:   string(bytes.TrimSpace(b[hicnStart:hicnEnd])),
-				//TODO: BeneficiaryID
-			}).Error
-			if err != nil {
-				return errors.Wrap(err, "could not create CCLF8 beneficiary record")
-			}
-		}
+		return err
 	}
 
-	fmt.Printf("Imported CCLF8 file %s.\n", fileMetadata.name)
-	log.Infof("Imported CCLF8 file %s.", fileMetadata.name)
-
 	return nil
 }
 
-func importCCLF9(fileMetadata cclfFileMetadata) error {
-	fmt.Printf("Importing CCLF9 file %s...\n", fileMetadata.name)
-	log.Infof("Importing CCLF9 file %s...", fileMetadata.name)
+func importCCLF(fileMetadata cclfFileMetadata, importFunc func(uint, []byte, *gorm.DB) error) error {
+	fmt.Printf("Importing CCLF%d file %s...\n", fileMetadata.cclfNum, fileMetadata.name)
+	log.Infof("Importing CCLF%d file %s...", fileMetadata.cclfNum, fileMetadata.name)
 
 	if (cclfFileMetadata{}) == fileMetadata {
-		err := errors.New("file CCLF9 not found")
+		err := errors.New("CCLF file not found")
 		log.Error(err)
 		return err
 	}
 
-	if fileMetadata.cclfNum != 9 {
-		err := fmt.Errorf("expected CCLF file number 9, but was %d", fileMetadata.cclfNum)
+	r, err := zip.OpenReader(filepath.Clean(fileMetadata.filePath))
+	if err != nil {
+		err := errors.Wrapf(err, "could not read CCLF%d archive %s", fileMetadata.cclfNum, fileMetadata.name)
 		log.Error(err)
 		return err
 	}
+	defer r.Close()
```

Comment:
should this be above the error handling?  If we get an error r won't be closed

Reply:
I think the Go convention is to return nil for the thing being created if there's an error, so there wouldn't be a reader we could close. Looks to be how they handle it here: https://golang.org/src/archive/zip/reader.go?s=911:960#L40